doc/psud/PSU_daemon_design.md

```diff
@@ -33,37 +44,95 @@ PSU daemon data collection flow diagram:
 
 Now psud collects PSU data via platform API, and it also support platform plugin for backward compatible. All PSU data will be saved to redis database for further usage.
 
+### 2.1 PSU data collection specific to PSU power exceeding check
+
+We will leverage the existing framework of PSU daemon, adding corresponding logic to perform PSU power check.
+
+Currently, PSU daemon is waken up periodically, executing the following flows (flows in bold are newly introduced by the feature):
+
+1. Check the PSUs' physical entity information and update them into database
+2. Check the PSUs' present and power good information and update them to database
+   - __It will check the capability of PSU power check via reading PSU power thresholds when a new PSU is detected.__
+3. Check and update the PSUs' data
+   - Fetch voltage, current, power via calling platform API
+   - __Perform PSU power checking logic__
+   - Update all the information to database
+
+We will detail the new flows in the following sections.
+
+#### New PSU is detected
+
+Basically, there are two scenarios in which a new PSU can be detected:
+
+- On PSU daemon starting, all PSUs installed on the switch are detected
+- On new PSU pulgged, the new PSU is detected
+
+When one or more new PSUs is detected and power is good, PSU daemon tries retrieving the warning and critical threshold for each PSU installed on the switch.
+
+The PSU power checking will not be checked for a PSU if `NotImplemented` exception is thrown or `None` is returned while either threshold is being retrieved
+
+#### Alarm raising and clearing threshold
+
+We use asymmetric thresholds between raising and clearing the alarm for the purpose of creating a hysteresis and avoiding alarm flapping.
+
+- an alarm will be raised when a PSU's power is rising accross the critical threshold
+- an alarm will be cleared when a PSU's power is dropping across the warning threshold
+
+#### PSU power checking logic
+
+For each PSU supporting power checking:
+
+1. Retrieve the current power
+2. If flag `PSU power exceeded threshold` is `true`, compare the current power against the warning threshold
+   - If `current power` < `warning threshold`
+     - Set `PSU power exceeded threshold` to `false`
+     - Message in NOTICE level should be logged: `PSU <x>: current power <power> is below the warning threshold <threshold>` where
+       - `<x>` is the number of the PSU
+       - `<power>` is the current power of the PSU
+       - `<threshold>` is the warning threshold of the PSU
+   - Otherwise: no action
+3. Otherwise, compare the current power against the critical threshold
+   - If `current power` >= `critical threshold`
+     - Set `PSU power exceeded threshold` to `true`
+     - Message in WARNING level should be logged: `PSU <x>: current power <power> is exceeding the critical threshold <threshold>` where
```

Comment:
Should we clarify to the user that PSU redundancy is not supported?

Reply:
I'm not sure whether we should put it here. this is a generic feature to check PSU against the threshold. For us, the purpose is to notify users of losing redundancy but other vendors may have other purposes.